mllib/src/main/scala/org/apache/spark/ml/feature/VectorSlicer.scala

```diff
@@ -0,0 +1,153 @@
+
+/**
+ * :: Experimental ::
+ * Given either indices or names, it takes a vector column and output a vector column with the
+ * specified subset of features. Note that the vector column should contain ML [[Attribute]].
+ */
+@Experimental
+final class VectorSlicer(override val uid: String)
+  extends Transformer with HasInputCol with HasOutputCol {
+
+  def this() = this(Identifiable.randomUID("vectorSlicer"))
+
+  /**
+   * An array of indices to select features from a vector column.
+   * @group param
+   */
+  val selectedIndices = new IntArrayParam(this, "selectedIndices",
+    "An array of indices to select features from a vector column",
+    (x: Array[Int]) => if (x.isEmpty) true else x.min >= 0)
+
+  setDefault(selectedIndices -> Array.empty[Int])
+
+  /** @group getParam */
+  def getSelectedIndices: Array[Int] = getOrDefault(selectedIndices)
+
+  /** @group setParam */
+  def setSelectedIndices(value: Array[Int]): this.type = set(selectedIndices, value)
+
+  /**
+   * An array of feature names to select features from a vector column.
+   * @group param
+   */
+  val selectedNames = new StringArrayParam(this, "selectedNames",
+    "An array of feature names to select features from a vector column")
+
+  setDefault(selectedNames -> Array.empty[String])
+
+  /** @group getParam */
+  def getSelectedNames: Array[String] = getOrDefault(selectedNames)
+
+  /** @group setParam */
+  def setSelectedNames(value: Array[String]): this.type = set(selectedNames, value)
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  /**
+   * Slice a dense vector with an array of indices.
+   */
+  private def selectColumns(indices: Array[Int], features: DenseVector): Vector = {
+    Vectors.dense(indices.map(features.apply))
+  }
+
+  /**
+   * Slice a sparse vector with a set of indices.
+   */
+  private def selectColumns(indices: Set[Int], features: SparseVector): Vector = {
+    Vectors.sparse(
+      indices.size, features.indices.zip(features.values).filter(x => indices.contains(x._1)))
+  }
+
+  private def getFeatureIndicesFromNames(inputAttr: AttributeGroup): Array[Int] = {
+    $(selectedNames).map(name => inputAttr.getAttr(name).index.get)
+  }
+
+  private def merge(xs: List[Int], ys: List[Int]): List[Int] = {
+    (xs, ys) match {
+      case (Nil, ys) => ys
+      case (xs, Nil) => xs
+      case (x :: xs1, y :: ys1) =>
+        if (x < y) x :: merge(xs1, ys)
+        else y :: merge(xs, ys1)
+    }
+  }
+
+  /**
+   * Union feature indices from user specified indices and indices that transformed from user
+   * specified feature names. After the union process, indices are sorted ASC and any potential
+   * duplicates are removed.
+   */
+  private def unionFeatureIndices(first: Array[Int], second: Array[Int]): Array[Int] = {
+    merge(first.sorted.toList, second.sorted.toList).distinct.toArray
```

Comment:
I would prefer to follow the order given by the user.  Users who specify features by a list of names may later try to index the resulting vector using the same order as that list.
This should be easy to implement if we ensure the selected indices and names do not overlap beforehand.